### PR TITLE
Fix issue #206 storing Hilbert as shared_ptr in Operator

### PR DESCRIPTION
**Pull request by gcarleo**
I wanted to do this using std::unique_ptr instead, but I had some issues with deleted copy constructors etc... maybe shared_ptr is good enough

### CODE REVIEW

**Comment by PhilipVinc** [comment]
Please see my last comment on #206 . I think this is the good solution (we should also do the same for graphs, BTW) but I would do the following change:

 - Store the `shared_ptr<AbstractHilbert>` as a public object and use the ptr in lieu of `GetHilbert` more or less everywhere.
 - Change functions that take an `AbstractHilbert&` to take a `shared_ptr<AbstractHilbert>` (or an rvalue `&&` where appropriate)
 - Update the python bindings to store `Hilbert` with a `shared_ptr` and not by value.

This is because the dereference operation in 
```
AbstractHilbert& GetHilbert() {return *hilbert;}
```

is not tracked by the `shared_ptr`.

(BTW, this is also an issue in my last PR).

**Comment by gcarleo** [comment]
> This is because the dereference operation in
> 
> AbstractHilbert& GetHilbert() {return *hilbert;}
> is not tracked by the shared_ptr.

I am not sure that the solution is moving everything to shared_ptr in C++ though. I tried to hide shared_ptr usage (that's why `GetHilbert` and `SetHilbert` are there), because I believe that  using `shared_ptr` is not a good coding practice in C++. Unfortunately we need it mostly because of python bindings, but otherwise I wouldn't make it a primary type on the C++ side... 

As you say, one solution is to make `GetHilbert` return a const r-value, and define a `GetHilbertCopy` returning a clone of Hilbert that can be stored as shared_ptr.  Still I am not entirely sure that the current design is problematic/ can generate bugs on the python side. Do you have an example where we can have a bug? 

My point is that when we pass `Hilbert` to `Operator` , `Operator` makes a clone and stores a clone of `Hilbert`, not the original pointer. So when you do `GetHilbert`, the clone is returned. As far as I understand this is completely fine, because pybind will keep alive the attributes of a class.

**Comment by femtobit** [comment]
> I am not sure that the solution is moving everything to shared_ptr in C++ though. I tried to hide shared_ptr usage (that's why GetHilbert and SetHilbert are there), because I believe that using shared_ptr is not a good coding practice in C++.

While using `shared_ptr` *unnecessarily* can be considered bad practice in C++, this is not what we would be doing. Most of code is already written under the assumption that the Hilbert space stays alive as long as objects referencing it exists (an assumption that, as we have seen, fails when the Hilbert space is created from Python). Thus, having shared ownership of the Hilbert space by the objects referencing it makes sense and using `shared_ptr` to make this explicit [is certainly good practice](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-shared_ptr). (And going an extra mile to use `shared_ptr` but then hide that fact is arguably more confusing and potentially error-prone than just being explicit about it.)

I also do not see a need to clone the Hilbert space (this kind of defeats the purpose of using `shared_ptr` in the first place).

I thus very much agree with the approach that @PhilipVinc outlined [in his preceding comment](https://github.com/netket/netket/pull/207#issuecomment-500209375).

**Comment by gcarleo** [comment]
> I thus very much agree with the approach that @PhilipVinc outlined in his preceding comment.

Again, can you provide here an example where the currently implemented approach of this PR breaks?

**Comment by femtobit** [comment]
Not sure whether it breaks; but copying (or `Clone`ing) the Hilbert space defeats the purpose of having `shared_ptr`s, i.e., having one Hilbert object with shared ownership between the operators and machines using it.

If you want each operator (and machine) should own its own copy of `Hilbert`, then there is not need for `shared_ptr`. (And since we did not copy around the Hilbert object before, why start now?)

**Comment by gcarleo** [comment]
My main problem with `shared_ptr` is that, despite being a concrete object, you still need to use the -> syntax to call methods of the bound class. `->` is just ugly, and one needs to change the code basically everywhere there is a call to `Hilbert`... it's fine, it can be done, it's just a pain...

**Comment by femtobit** [comment]
I did the changes to use `shared_ptr` consistently in #208.

> My main problem with shared_ptr is that, despite being a concrete object, you still need to use the -> syntax to call methods of the bound class. -> is just ugly, and one needs to change the code basically everywhere there is a call to Hilbert... it's fine, it can be done, it's just a pain...

I can understand the dislike for `->`, but getting rid of it at the cost of introducing copies of the Hilbert space seems like overkill to me.

In my commit above, `AbstractOperator::GetHilbert` still returns a reference (so you don't need to use `->` everywhere :slightly_smiling_face:), which is fine as long as the reference is not used after the end of the operator's lifetime. (And via [`GetHilbert().shared_from_this()`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this) it can be converted to a `shared_ptr` when needed.)

**Comment by PhilipVinc** [comment]
Thanks @femtobit , you said exactly what I think on `shared_ptr`.

@gcarleo I tried to play around a bit and I can't cause a segmentation fault with this PR, but that is not the point: my interpretation of modern C++ is that the compiler should correctly compile only code that won't lead to runtime errors, and when you define a new class you should correctly define all methods to allow the compiler to do so.

Here you are breaking this contract: by returning a reference from `GetHilbert` you are allowing the compiler to generate valid code even when this code would break. You implicitly require a programmer to call `::Clone()` whenever he keeps a reference to the `Hilbert` object. 

In your PR you always call `Clone()` when needed, but in the future, if someone submits a PR and does not know this, he might introduce a bug. So preventing this code from compiling alltogether is what (I believe) should be done.

File: Sources/Machine/DensityMatrices/abstract_density_matrix.hpp [hunk 1]
```diff
@@ -78,25 +75,38 @@ class AbstractDensityMatrix : public AbstractMachine {
...
+  AbstractDensityMatrix(std::shared_ptr<const AbstractHilbert> physical_hilbert,
+                        std::unique_ptr<const AbstractGraph> doubled_graph)
+      : AbstractMachine(std::make_shared<CustomHilbert>(
+            *doubled_graph, physical_hilbert->LocalStates())),
+        hilbert_physical_(physical_hilbert),
+        graph_doubled_(std::move(doubled_graph)) {}
+
+  AbstractDensityMatrix(std::shared_ptr<const AbstractHilbert> physical_hilbert,
+                        std::shared_ptr<const AbstractHilbert> doubled_hilbert)
+      : AbstractMachine(doubled_hilbert), hilbert_physical_(physical_hilbert) {}
```
Comment (twesterhout) [inline Sources/Machine/DensityMatrices/abstract_density_matrix.hpp]: One of these two protected constructors should probably be removed.

File: Sources/Machine/DensityMatrices/abstract_density_matrix.hpp [hunk 2]
```diff
@@ -78,25 +75,38 @@ class AbstractDensityMatrix : public AbstractMachine {
...
-  const AbstractHilbert &GetHilbertPhysical() const noexcept {
+  std::shared_ptr<const AbstractHilbert> GetHilbertPhysicalShared() const {
     return hilbert_physical_;
   }
+
+  /* Member function returning a reference to the physical hilbert space
+   * on which this density matrix acts.
+   * @return The physical hilbert space
+   */
+  const AbstractHilbert &GetHilbertPhysical() const {
```
Comment (twesterhout) [inline Sources/Machine/DensityMatrices/abstract_density_matrix.hpp]: `noexcept`?

**Comment by gcarleo** [comment]
@twesterhout can we merge?

**Pull request merged**